### PR TITLE
Add a date setting

### PR DESCRIPTION
**Pull request by robert00s**
Close: #2265 
See: https://github.com/koreader/koreader/pull/3085#issuecomment-322291560
It's work in emulator and Kindle.

screenshot 2017-09-17 19-58-20

### CODE REVIEW

File: frontend/device/kindle/device.lua [hunk 1]
```diff
@@ -66,13 +66,22 @@ function Kindle:supportsScreensaver()
     end
 end
 
-function Kindle:setTime(hour, min)
+function Kindle:setTime(year, month, day, hour, min, sec)
```
Comment (Frenzie) [inline frontend/device/kindle/device.lua]: I'd say that it should be renamed to setDate with the new functionality.

Possibly with a convenience front like
```lua
function setTime(hour, min, sec)
  self:setDate(nil, nil, nil, hour, min, sec)
end
```

File: frontend/device/kindle/device.lua [hunk 2]
```diff
@@ -66,13 +66,22 @@ function Kindle:supportsScreensaver()
...
     if hour == nil or min == nil then return true end
-    if os.execute(string.format("date -s '%d:%d'", hour, min)) == 0 then
-        os.execute('hwclock -u -w')
-        return true
+    if year then
```
Comment (Frenzie) [inline frontend/device/kindle/device.lua]: Should that be `if year and month`?

File: frontend/device/kindle/device.lua [hunk 3]
```diff
@@ -66,13 +66,22 @@ function Kindle:supportsScreensaver()
...
+        if os.execute(string.format("date -s '%d-%d-%d %d:%d:%d'", year, month, day, hour, min, sec)) == 0 then
```
Comment (Frenzie) [inline frontend/device/kindle/device.lua]: How about something like this?

```lua
local command
if year and month then
command = string.format("date -s '%d-%d-%d %d:%d:%d'", year, month, day, hour, min, sec)
else
command = string.format("date -s '%d:%d'",hour, min)
end
if os.execute(command) == 0 then
  os.execute(hwclock)
  return true
end
return false
```

File: frontend/ui/elements/common_settings_menu_table.lua [hunk 4]
```diff
@@ -22,32 +23,69 @@ end
...
-                            text = T(_("Current time: %1:%2"), string.format("%02d", now_t.hour),
-                                string.format("%02d", now_t.min))
-                        })
-                    else
-                        UIManager:show(InfoMessage:new{
-                            text = _("Time couldn't be set"),
-                        })
-                    end
-                end
+        text = _("Set time and date"),
```
Comment (Frenzie) [inline frontend/ui/elements/common_settings_menu_table.lua]: Just `Time and date` will do. :-)

File: frontend/ui/elements/common_settings_menu_table.lua [hunk 5]
```diff
@@ -22,32 +23,69 @@ end
...
+                                })
+                            end
+                        end
+                    }
+                    UIManager:show(date_widget)
+                end,
             }
-            UIManager:show(time_widget)
-        end,
+
```
Comment (Frenzie) [inline frontend/ui/elements/common_settings_menu_table.lua]: stray newline

File: frontend/ui/widget/numberpickerwidget.lua [hunk 6]
```diff
@@ -142,14 +157,32 @@ function NumberPickerWidget:changeValue(value, step, max, min)
...
+            value = max
         end
     end
     return value
 end
 
+function NumberPickerWidget:getDaysInMonth(month, year)
+    local days_in_month = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 }
+    local days = days_in_month[month]
+    -- check for leap year
```
Comment (Frenzie) [inline frontend/ui/widget/numberpickerwidget.lua]: Hah, you made me look back at something I wrote over 14 years ago ([link](http://fransdejonge.com/calendar/)).

```javascript
// makes sure months have the right length...
var leapdays = new Array(31,29,31, 30,31,30, 31,31,30, 31,30,31); 
var yeardays = new Array(31,28,31, 30,31,30, 31,31,30, 31,30,31); 

// […]

function isLeapYear( year ) {
  // is it leap year ? returns a boolean
  return ( (0 == (year%4)) && ( (0 != (year%100)) || (0 == (year%400))));
  // ie, if the year divides by 4, but not by 100 except when it divides by
  // 400, it is leap year
}
```

Comment (robert00s) [reply frontend/ui/widget/numberpickerwidget.lua]: ;-)

File: frontend/device/kindle/device.lua [hunk 7]
```diff
@@ -66,9 +66,15 @@ function Kindle:supportsScreensaver()
     end
 end
 
-function Kindle:setTime(hour, min)
+function Kindle:setDateTime(year, month, day, hour, min, sec)
```
Comment (Frenzie) [inline frontend/device/kindle/device.lua]: setDateTime, nice one ;-D

**Comment by Frenzie** [comment]
Feel free to merge once the check passes. You should see it here?

screenshot_2017-09-18_17-40-14

**Comment by robert00s** [comment]
Unfortunately....
screenshot 2017-09-18 18-15-54

**Comment by Frenzie** [comment]
Oh! I'd added you to the Kindle dev team, but apparently only the Kobo dev team had the relevant permissions. It should work now.

**Comment by robert00s** [comment]
Still the same :)
I try reload my session but no effect.

**Comment by Frenzie** [comment]
My bad, by default adding a repository to a team gives you read access. Which doesn't really mean anything on a public repository. :-P

**Comment by robert00s** [comment]
Thanks, now it works :+1:

**Comment by Frenzie** [comment]
Good. :)

**Pull request merged**